components/locale_canonicalizer/src/locale_canonicalizer.rs

```diff
@@ -0,0 +1,164 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/master/LICENSE ).
+use crate::error::LocaleCanonicalizerError;
+use icu_locid::LanguageIdentifier;
+use icu_locid::Locale;
+use icu_provider::prelude::*;
+use icu_provider::structs::likelysubtags::*;
+use std::borrow::Cow;
+
+pub struct LocaleCanonicalizer<'a> {
+    provider: &'a dyn DataProvider<'a>,
+    likely_subtags: Option<LikelySubtagsV1>,
+}
+
+impl LocaleCanonicalizer<'_> {
+    fn ensure_likely_subtags(&mut self) -> Result<&LikelySubtagsV1, DataError> {
+        if self.likely_subtags.is_none() {
+            let payload: Cow<LikelySubtagsV1> = self
+                .provider
+                .load(&DataRequest {
+                    data_key: key::LIKELY_SUBTAGS_V1,
+                    data_entry: DataEntry {
+                        variant: None,
+                        langid: "und".parse().unwrap(),
+                    },
+                })?
+                .take_payload()?;
+            let mut likely_subtags = payload.into_owned();
+            likely_subtags.entries.sort();
+            self.likely_subtags = Some(likely_subtags);
+        }
+        Ok(&self.likely_subtags.as_ref().unwrap())
+    }
+
+    /// A constructor which takes a DataProvider and creates a
+    /// LocaleCanonicalizer.
+    pub fn new<'a>(provider: &'a dyn DataProvider<'a>) -> LocaleCanonicalizer<'a> {
+        LocaleCanonicalizer {
+            provider,
+            likely_subtags: None,
+        }
+    }
+
+    /// This returns a new Locale that is the result of running the
+    /// 'Add Likely Subtags' algorithm from
+    /// https://www.unicode.org/reports/tr35/#Likely_Subtags.
+    pub fn maximize(&mut self, locale: &Locale) -> Result<Locale, LocaleCanonicalizerError> {
+        let likely_subtags = self.ensure_likely_subtags()?;
+
+        let mut key = LanguageIdentifier {
+            language: locale.language,
+            script: locale.script,
+            region: locale.region,
+            variants: icu_locid::subtags::Variants::from_vec_unchecked(Vec::new()),
+        };
```

Comment:
I tried the LI From for Locale, but because we have a &Locale here and From takes a Locale, I thought it ended up being uglier than just constructing the LI directly.

Reply:
The unfortunate consequence of that situation is that you can't just pass `LanguageIdentifier` here then.

I'm starting to lean toward an idea that we want to have a trait `LanguageIdentifierType` that is implemented by `LanguageIdentifier` and `Locale` and accept either here.

That would be a non-trivial change unfortunately.

@sffc @Manishearth - thoughts?

Reply:
Another option is that if Locale wraps a LanguageIdentifier (as a field), then we could `impl AsRef<LanguageIdentifier> for Locale`

Reply:
> I tried the LI From for Locale, but because we have a &Locale here and From takes a Locale, I thought it ended up being uglier than just constructing the LI directly.

You could also add `From<&Locale> for LanguageIdentifier`, but it would clone in `From`.

Reply:
Another observation: since you set the variants to `subtags::Variants::default()` and never use them for maximizing subtags, you could make an LSR type for this use case and use it instead of LangugeIdentifier.  This approach of making use-case-specific versions of LanguageIdentifier was agreed upon in #243.

Reply:
Locale wrapping LI with a `.id()` method (and perhaps some trait impls) makes sense to me

Reply:
Personally, I like the idea of having Locale wrap LanguageIdentifier. Are there any objections to me filing a separate issue to discuss that, as that seems like too big of a change to consider here?

Reply:
I'm in favor of that

Reply:
Filed https://github.com/unicode-org/icu4x/issues/447

Reply:
+1, Locale == LanguageIdentifier + something extra "feels right" to me.
---

BTW: since this is all about locale, we need to make sure it really works.

So can we add some test units?
Something like
```
zh-hant-u-nu-Chinese-hc-h24 => maximize => zh-Hant-TW-u-hc-h24-nu-chinese
zh-Hant-TW-u-hc-h24-nu-chinese => minimized => zh-TW-u-hc-h24-nu-chinese
```